### PR TITLE
Second assignment

### PR DESCRIPTION
**Pull request by dwgoblue**
Upload the assignment in which I wrote Cross validation, Lasso and Ridge regression.

### CODE REVIEW

File: DW_Assignment_Jul72020.ipynb [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+      "(20, 5)\n",
+      "(20, 5)\n",
+      "(20, 5)\n"
+     ]
+    }
+   ],
+   "source": [
+    "X = np.random.rand(100,5)\n",
+    "Y = np.random.rand(100,5)\n",
+    "integrate_split(X, Y)"
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:143]: I followed the logic until here - what is the `integrate_split()` function? is that your cross-validation function that splits into 5-folds?

File: DW_Assignment_Jul72020.ipynb [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 42,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Train and Test Data Split.\n",
+    "def trainTestSplit(X, Y, subset_size=0.8, seed=20):\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:43]: Perfect

File: DW_Assignment_Jul72020.ipynb [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+    "    \n",
+    "    def __init__(self, X, Y, LAMBDA=2, subset_size=0.8, alpha=0.05, epochs=100):\n",
+    "        # Initialize data\n",
+    "        self.beta = np.random.rand()\n",
+    "        self.intercept = np.random.rand()\n",
+    "        self.alpha = alpha\n",
+    "        self.epochs = epochs\n",
+    "        self.LAMBDA = LAMBDA\n",
+    "        self.mse = np.zeros(epochs)\n",
+    "        self.Lasso = np.zeros(epochs) #L1 Regularization\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:315]: Not sure if this is the right way to initialize your LASSO or ridge coefficient variable. The proper size will be the number of predictors / features in the dataset, not the number of epochs.

File: DW_Assignment_Jul72020.ipynb [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+    "        \n",
+    "        # Linear regression function\n",
+    "        def lm(X, beta, intercept):\n",
+    "            yhat = X * beta + intercept\n",
+    "            return yhat\n",
+    "        \n",
+    "        # Learning with regression\n",
+    "        for i in range(self.epochs):\n",
+    "            ypred= lm(self.dataset['Xtrain'], self.beta, self.intercept)\n",
+    "            dB1 = np.mean([(y1-y2) for y1, y2 in zip(ypred, self.dataset['Ytrain'])])\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:349]: This does work, however, I don't think you need the list comprehension. To simplify the expression, if you do np.mean(ypred - self.dataset['Ytrain']), I believe this will still perform element-wise subtraction followed by the np.mean() operation.

File: DW_Assignment_Jul72020.ipynb [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+    "            dB2 = np.mean([xi*(y1-y2) for y1, y2, xi in zip(ypred, self.dataset['Ytrain'], self.dataset['Xtrain'])])\n",
+    "            self.mse[i] = 0.5*np.sum([(y1-y2)**2 for y1, y2 in zip(ypred, self.dataset['Ytrain'])])\n",
+    "            self.Lasso[i] = self.mse[i] + self.LAMBDA*self.beta**2\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:352]: yup, this gives the regularized cost

File: DW_Assignment_Jul72020.ipynb [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+    "            self.Ridge[i] = self.mse[i] + abs(self.LAMBDA*self.beta)\n",
+    "            # Update...\n",
+    "            self.beta = self.beta - self.alpha*dB2\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:355]: So we didn't cover this, but regularized regression coefficients and the intercept term also have a penalized expression for gradient descent. So depending on L1 or L2 regularization, there will be an extra penalty term for your updates. This Medium article provides the expressions: https://towardsdatascience.com/intuitions-on-l1-and-l2-regularisation-235f2db4c261. Otherwise, the logic you wrote is correct.

File: DW_Assignment_Jul72020.ipynb [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+    "            self.intercept = self.intercept - self.alpha*dB1\n",
+    "        # Learning curve.\n",
+    "        plt.plot(self.Ridge)\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:358]: It's awesome you're showing the Ridge cost curve, and it looks like you got a good model! You don't need to do this part (because I think you got the high level concepts), but if you were to deploy this as an app or a software package, the next step would be to have it as an optional function argument for the user to plot out either lm, l1, or l2 cost curves.

File: DW_Assignment_Jul72020.ipynb [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+    "        plt.show()\n",
+    "\n",
+    "    def Prediction(self):\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:361]: Code looks great!

File: DW_Assignment_Jul72020.ipynb [hunk 9]
```diff
@@ -0,0 +1,624 @@
...
+    "        # Predict using model\n",
+    "        final_ypred = self.dataset['Xtest'] * self.beta + self.intercept\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:363]: It's nice that you provide some diagnostic plots with your predict function. The next steps would be to add axis labels to your plots.

File: DW_Assignment_Jul72020.ipynb [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+    "        # Evaluate model\n",
+    "        r = np.corrcoef(final_ypred.flatten(), self.dataset['Ytest'].flatten())\n",
+    "        plt.plot(final_ypred)\n",
+    "        plt.plot(self.dataset['Ytest'])\n",
+    "        plt.show()\n",
+    "        r2 = r**2\n",
+    "        mse = 0.5*np.sum([(y1-y2)**2 for y1, y2 in zip(final_ypred, self.dataset['Ytest'])])\n",
+    "        Lasso = mse + self.LAMBDA*self.beta**2\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:371]: These are the LASSO and Ridge costs, but not the predictions. If you want to get the predictions, you need to cache the coefficients from L1 and L2 regularization in the previous function. Then do the `final_ypred` computation with the L1 and L2 regression coefficient.

File: DW_Assignment_Jul72020.ipynb [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+    "    # Split the data into k folds.\n",
+    "    CVS_X = CrossValidationSplit(X, folds=folds)\n",
+    "    CVS_Y = CrossValidationSplit(Y, folds=folds)\n",
+    "    # Split data in each fold into test and training set.\n",
+    "    score = {'R2':0, 'L1':0, 'L2':0}\n",
+    "    for i in range(folds):\n",
+    "        m = model(CVS_X[i], CVS_Y[i], alpha=10**(-3))\n",
+    "        m.RegressionLearning()\n",
+    "        performance = m.Prediction()\n",
+    "        score['R2'] += performance['r2-value']\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:439]: So this does not output a cross-validation person correlation coefficient score or model. Looking at this code, it looks like the `+=` operation is performing an update by addition. Thus, you're adding your values the R2, and the L1 and L2 MSEs, which doesn't really tell us much about the model performance from cross validation.

If I wanted to select the best model from cross validation, I would do the following steps in the for loop:
  * Train an OLS, L1, and/or L2 model, and cache those coefficients in a dict where the key is the kfold  and the value is the coefficients. 
  * Output a prediction for OLS, L1, and L2
  * Compute the R or R2 value for each k-fold and cache it 

Out of the for loop:
  * From the cached R or R2, get the key for the highest value
  * Use that key to get your model coefficients. This is the best model from k-fold cross validation with the highest R or R2.

Ideally, you want to perform a hold out initially, and with the bigger dataset, use that for kfold CV to get the best model. Then, with the smaller dataset, use that best model to report the final R/R2/MSE value. This would be what you report in a publication.

File: DW_Assignment_Jul72020.ipynb [hunk 12]
```diff
@@ -0,0 +1,624 @@
+{
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:1]: Great job Da-Wei! I hope you found the exercise useful in understanding how to implement regression from scratch, and what it does underneath the hood. Hopefully, you'll feel more comfortable when the time comes where you'll be using regression for your own analyses. 

You got many of the high level concepts correct, and I offer suggestions/solutions and an explanation on where things can be improved or corrected.

Feel free to either respond to the comments or DM me if you have questions. Once you finish correcting some parts of the code, and we do one final review, you can a) add your name to the notebook to indicate you're the primary author, and b) push this to the master branch.

File: DW_Assignment_Jul72020.ipynb [hunk 13]
```diff
@@ -0,0 +1,624 @@
...
+    "        # Using these shuffled indicies, get the dataset.\n",
+    "        Xtrain = X[trainIdx, :]\n",
+    "        Ytrain = Y[trainIdx, :]\n",
+    "        Xtest = X[testIdx, :]\n",
+    "        Ytest = Y[testIdx, :]\n",
+    "        # Updata dataset in class property.\n",
+    "        return {'Xtrain':Xtrain, 'Ytrain':Ytrain, 'Xtest':Xtest, 'Ytest':Ytest}\n",
+    "    \n",
+    "\n",
+    "    def RegressionLearning(self):\n",
```
Comment (ScottCampit) [inline DW_Assignment_Jul72020.ipynb:339]: My suggestion would be to offer the user an option to specify whether they want to run least squares, L1 or L2 regularization when calling the object.
